File: sdk/typescript/test/client.test.mjs
// Tests for the TypeScript kern SDK client.
// Uses Node's built-in test runner + a mocked global fetch — no real HTTP calls.
// Imports the compiled output from dist/, so run `npm run build` first.

import { test } from "node:test";
import assert from "node:assert/strict";
import { spawn, spawnSync } from "node:child_process";
import { existsSync, openSync, closeSync, readFileSync } from "node:fs";
import net from "node:net";
import os from "node:os";
import path from "node:path";
import { fileURLToPath } from "node:url";
import { Client, KernError } from "../dist/index.js";

// The REAL fetch is captured at module load: the mocked tests below replace
// global.fetch and never restore it, so the live cross-check must use the
// original (see the live test at the bottom of this file).
const realFetch = globalThis.fetch.bind(globalThis);

// loadContractFixture reads the SHARED contract fixture every SDK suite
// asserts against (sdk/contract/tools_call.json — the POST /v1/tools/{name}
// route shape). A route-shape change must fail the Go, Python and TypeScript
// suites until the clients are updated.
function loadContractFixture() {
  const url = new URL("../../contract/tools_call.json", import.meta.url);
  const fx = JSON.parse(readFileSync(url, "utf8"));
  assert.equal(fx._version, 1, "contract fixture version must be 1");
  assert.ok(fx.cases.length > 0, "contract fixture must have cases");
  return fx.cases;
}

// Helper: capture the fetch call arguments.
function mockFetch(responseBody, status = 200, statusText = "OK") {
  const calls = [];
  global.fetch = async (url, init) => {
    calls.push({ url, init });
    return {
      ok: status >= 200 && status < 300,
      status,
      statusText,
      text: async () => JSON.stringify(responseBody),
    };
  };
  return calls;
}

test("analyze posts change to /v1/analyze", async () => {
  const calls = mockFetch({ packet: {}, text: "ok" });
  const c = new Client("http://test:8090/");
  const out = await c.analyze("change x");
  assert.equal(out.text, "ok");
  assert.equal(calls[0].url, "http://test:8090/v1/analyze");
  assert.equal(calls[0].init.method, "POST");
  assert.equal(JSON.parse(calls[0].init.body).change, "change x");
  assert.equal(calls[0].init.headers["Content-Type"], "application/json");
});

test("plan posts change to /v1/plan", async () => {
  const calls = mockFetch({ packet: {}, text: "plan" });
  const c = new Client("http://test:8090");
  await c.plan("change y");
  assert.equal(calls[0].url, "http://test:8090/v1/plan");
  assert.equal(JSON.parse(calls[0].init.body).change, "change y");
});

test("whatIf builds body with optional new_target", async () => {
  const c = new Client("http://test:8090");
  let calls = mockFetch({});
  await c.whatIf("change", "log");
  assert.deepEqual(JSON.parse(calls[0].init.body), { change: "change", kind: "log" });

  calls = mockFetch({});
  await c.whatIf("change", "log", "http://new");
  assert.deepEqual(JSON.parse(calls[0].init.body), {
    change: "change",
    kind: "log",
    new_target: "http://new",
  });
});

test("memoryAdd posts content/type/scope/tags to /v1/memory", async () => {
  const calls = mockFetch({});
  const c = new Client("http://test:8090");
  await c.memoryAdd("content", "note", "svc", ["a", "b"]);
  assert.equal(calls[0].url, "http://test:8090/v1/memory");
  assert.deepEqual(JSON.parse(calls[0].init.body), {
    content: "content",
    type: "note",
    scope: "svc",
    tags: ["a", "b"],
  });
});

test("memoryAdd uses defaults", async () => {
  const calls = mockFetch({});
  const c = new Client("http://test:8090");
  await c.memoryAdd("content");
  assert.deepEqual(JSON.parse(calls[0].init.body), {
    content: "content",
    type: "lesson",
    scope: "",
    tags: [],
  });
});

test("health GETs /api/health", async () => {
  const calls = mockFetch({ status: "ok" });
  const c = new Client("http://test:8090");
  const out = await c.health();
  assert.deepEqual(out, { status: "ok" });
  assert.equal(calls[0].url, "http://test:8090/api/health");
  assert.equal(calls[0].init.method, "GET");
});

test("task GETs encoded id from /v1/tasks/{id}", async () => {
  const calls = mockFetch({});
  const c = new Client("http://test:8090");
  await c.task("abc def");
  assert.equal(calls[0].url, "http://test:8090/v1/tasks/abc%20def");
  assert.equal(calls[0].init.method, "GET");
});

test("loop posts intent and level to /v1/loop", async () => {
  const c = new Client("http://test:8090");
  let calls = mockFetch({});
  await c.loop("intent", "L3");
  assert.deepEqual(JSON.parse(calls[0].init.body), { intent: "intent", level: "L3" });

  calls = mockFetch({});
  await c.loop("intent");
  assert.deepEqual(JSON.parse(calls[0].init.body), { intent: "intent", level: "L0" });
});

test("execute POSTs patch to /v1/execute", async () => {
  const calls = mockFetch({});
  const c = new Client("http://test:8090");
  await c.execute("patch text");
  assert.equal(calls[0].url, "http://test:8090/v1/execute");
  assert.deepEqual(JSON.parse(calls[0].init.body), { patch: "patch text" });
});

test("audit GETs /v1/audit/{id} when a task id is given", async () => {
  const c = new Client("http://test:8090");
  const calls = mockFetch({});
  await c.audit("task-123");
  assert.equal(calls[0].url, "http://test:8090/v1/audit/task-123");
  assert.equal(calls[0].init.method, "GET");
});

test("audit throws when no task id is given", async () => {
  const c = new Client("http://test:8090");
  assert.throws(() => c.audit(""), /requires a taskId/);
  assert.throws(() => c.audit(undefined), /requires a taskId/);
});

test("non-2xx raises KernError with status", async () => {
  global.fetch = async () => ({
    ok: false,
    status: 404,
    statusText: "Not Found",
    text: async () => '{"error": "no such task"}',
  });
  const c = new Client("http://test:8090");
  await assert.rejects(
    () => c.task("missing"),
    (e) => e instanceof KernError && e.status === 404
  );
});

test("connection failure raises KernError", async () => {
  global.fetch = async () => {
    throw new Error("fetch failed");
  };
  const c = new Client("http://test:8090");
  await assert.rejects(
    () => c.memoryList(),
    (e) => e instanceof KernError && /connection error/.test(e.message)
  );
});
test("approvalsPending gets /v1/approvals/pending", async () => {
  const calls = mockFetch({ items: [] });
  const c = new Client("http://test:8090");
  const out = await c.approvalsPending();
  assert.deepEqual(out, { items: [] });
  assert.equal(calls[0].url, "http://test:8090/v1/approvals/pending");
  assert.equal(calls[0].init.method, "GET");
});

test("incidents gets /v1/incidents", async () => {
  const calls = mockFetch({ items: [] });
  const c = new Client("http://test:8090");
  const out = await c.incidents();
  assert.deepEqual(out, { items: [] });
  assert.equal(calls[0].url, "http://test:8090/v1/incidents");
});

test("callTool posts args to /v1/tools/{name} and returns the output payload", async () => {
  const calls = mockFetch({ output: "masked 1 secrets" });
  const c = new Client("http://test:8090");
  const out = await c.callTool("kern_mask_pii", { text: "token=sk-abc123" });
  assert.equal(out, "masked 1 secrets");
  assert.equal(calls[0].url, "http://test:8090/v1/tools/kern_mask_pii");
  assert.equal(calls[0].init.method, "POST");
  assert.deepEqual(JSON.parse(calls[0].init.body), { text: "token=sk-abc123" });
  assert.equal(calls[0].init.headers["Content-Type"], "application/json");
});

test("callTool defaults args to an empty object", async () => {
  const calls = mockFetch({ output: "" });
  const c = new Client("http://test:8090");
  const out = await c.callTool("kern_no_such_tool");
  assert.equal(out, "");
  assert.equal(calls[0].url, "http://test:8090/v1/tools/kern_no_such_tool");
  assert.deepEqual(JSON.parse(calls[0].init.body), {});
});

test("callTool obeys the shared contract fixture (sdk/contract/tools_call.json)", async (t) => {
  for (const tc of loadContractFixture()) {
    await t.test(tc.id, async () => {
      const status = tc.response.status;
      const body = tc.response.body;
      const url = `http://test:8090/v1/tools/${encodeURIComponent(tc.name)}`;
      if (status === 200) {
        const calls = mockFetch(body);
        const c = new Client("http://test:8090");
        const out = await c.callTool(tc.name, tc.args);
        assert.equal(out, tc.expect_output);
        assert.equal(calls[0].url, url);
        assert.equal(calls[0].init.method, "POST");
        assert.deepEqual(JSON.parse(calls[0].init.body), tc.args);
        assert.equal(calls[0].init.headers["Content-Type"], "application/json");
      } else {
        global.fetch = async () => ({
          ok: false,
          status,
          statusText: "err",
          text: async () => JSON.stringify(body),
        });
        const c = new Client("http://test:8090");
        await assert.rejects(
          () => c.callTool(tc.name, tc.args),
          (e) => e instanceof KernError && e.status === status,
        );
      }
    });
  }
});

test("incident gets /v1/incidents/{id} and requires an id", async () => {
  const calls = mockFetch({ ID: "inc-1" });
  const c = new Client("http://test:8090");
  const out = await c.incident("inc-1");
  assert.equal(out.ID, "inc-1");
  assert.equal(calls[0].url, "http://test:8090/v1/incidents/inc-1");
  assert.throws(() => c.incident(""), (e) => e instanceof KernError && /incidentId/.test(e.message));
});

test("eventsStream yields parsed SSE data payloads", async () => {
  const frames = 'event: ping\ndata: {"kind":"x","n":1}\n\ndata: plain-text\n\n';
  const stream = new ReadableStream({
    start(controller) {
      controller.enqueue(new TextEncoder().encode(frames));
      controller.close();
    },
  });
  global.fetch = async () => ({
    ok: true,
    status: 200,
    statusText: "OK",
    body: stream,
    text: async () => "",
  });
  const c = new Client("http://test:8090");
  const out = [];
  for await (const p of c.eventsStream()) out.push(p);
  assert.deepEqual(out, [{ kind: "x", n: 1 }, "plain-text"]);
});

test("eventsStream yields SSE data payloads (F-032 parity)", async () => {
  const stream = new ReadableStream({
    start(controller) {
      controller.enqueue(
        new TextEncoder().encode('data: {"id":1}\n\ndata: hello\n'),
      );
      controller.close();
    },
  });
  const calls = [];
  global.fetch = async (url, init) => {
    calls.push({ url, init });
    return { ok: true, status: 200, statusText: "OK", body: stream };
  };
  const c = new Client("http://test:8090", 10000);
  const out = [];
  for await (const p of c.eventsStream()) {
    out.push(p);
  }
  assert.equal(calls[0].url, "http://test:8090/v1/events/stream");
  assert.equal(calls[0].init.method, "GET");
  assert.equal(calls[0].init.headers.Accept, "text/event-stream");
  // JSON payloads decode; non-JSON payloads pass through as raw strings.
  assert.deepEqual(out, [{ id: 1 }, "hello"]);
});

test("eventsStream rejects non-2xx with KernError", async () => {
  global.fetch = async () => ({
    ok: false,
    status: 500,
    statusText: "boom",
    text: async () => "{}",
  });
  const c = new Client("http://test:8090", 10000);
  await assert.rejects(
    async () => {
      for await (const _ of c.eventsStream()) {
        /* consume */
      }
    },
    (e) => e instanceof KernError && e.status === 500,
  );
});

// --- OPTIONAL live-route cross-check (R4) ---
//
// The mocked tests above pin the shared fixture (sdk/contract/tools_call.json)
// against a fake transport; live-route drift detection rests on the Go suite's
// in-process route test. This test closes the gap: when a kern server can be
// started from the repo (a `go` toolchain + the repo root relative to this
// test file, falling back to a prebuilt `kern` binary on PATH), it starts
// `go run ./cmd/kern serve` on an ephemeral port with KERN_AUTH_TOKEN set and
// runs the fixture's 4 cases against the REAL POST /v1/tools/{name} route. It
// SKIPS gracefully (with a one-line reason) in CI without Go or in a
// standalone SDK checkout.

function kernRepoRoot() {
  // The test file lives in sdk/typescript/test/: the repo root is THREE
  // levels up (the contract fixture at sdk/contract is two, so the fixture
  // loader above uses "../..").
  return fileURLToPath(new URL("../../../", import.meta.url));
}

function findFreePort() {
  return new Promise((resolve, reject) => {
    const srv = net.createServer();
    srv.on("error", reject);
    srv.listen(0, "127.0.0.1", () => {
      const port = srv.address().port;
      srv.close(() => resolve(port));
    });
  });
}

function randomToken() {
  const bytes = new Uint8Array(16);
  crypto.getRandomValues(bytes);
  return Array.from(bytes, (b) => b.toString(16).padStart(2, "0")).join("");
}

function waitForKernServer(base, token, proc, logPath, timeoutMs, opts = {}) {
  const { usedPrebuilt = false, onSkip = () => {}, onStop = () => {} } = opts;
  const deadline = Date.now() + timeoutMs;
  return new Promise((resolve, reject) => {
    const poll = async () => {
      if (proc.exitCode !== null) {
        const tail = readFileSync(logPath, "utf8").split("\n").slice(-8).join("\n");
        reject(new Error(`kern serve exited early (code ${proc.exitCode}):\n${tail}`));
        return;
      }
      if (Date.now() > deadline) {
        reject(new Error(`kern serve did not become ready in ${timeoutMs}ms (slow first build? see ${logPath})`));
        return;
      }
      try {
        const resp = await realFetch(base + "/api/health", {
          headers: { Authorization: `Bearer ${token}` },
        });
        if (resp.ok) {
          if (usedPrebuilt) {
            // Prebuilt-binary fallback only: an installed `kern` may predate
            // the /v1/tools/{name} passthrough route (the route shipped with
            // the SDK-passthrough campaign). Probe it once; if the route is
            // absent, skip gracefully instead of failing the optional check
            // on an outdated binary.
            const probe = await realFetch(base + "/v1/tools/kern_no_such_tool", {
              method: "POST",
              headers: {
                "Content-Type": "application/json",
                Authorization: `Bearer ${token}`,
              },
              body: "{}",
            });
            const probeText = await probe.text();
            if (!probeText.includes('"error": "unknown tool')) {
              onStop();
              onSkip(
                "live cross-check skipped: prebuilt kern binary does not expose POST /v1/tools/{name} (too old?) — install a newer kern or use go run"
              );
              resolve(false); // skipped: test body returns early
              return;
            }
          }
          resolve(true);
          return;
        }
      } catch {
        /* not up yet — keep polling */
      }
      setTimeout(poll, 1000);
    };
    poll();
  });
}

test("live cross-check: fixture cases against a real kern serve process (skips without Go)", async (t) => {
  const repoRoot = kernRepoRoot();
  const goAvailable = spawnSync("go", ["version"], { stdio: "ignore" }).status === 0;
  const hasRepo = existsSync(path.join(repoRoot, "go.mod"));
  const kernOnPath =
    spawnSync("sh", ["-c", "command -v kern"], { stdio: "ignore" }).status === 0;
  let cmd;
  let usedPrebuilt = false;
  if (goAvailable && hasRepo) {
    cmd = ["go", "run", "./cmd/kern", "serve"]; // prefer repo go run (freshness)
  } else if (kernOnPath) {
    cmd = ["kern", "serve"];
    usedPrebuilt = true;
  } else {
    t.skip(
      "live cross-check skipped: no Go toolchain and no kern binary on PATH (CI without Go / standalone SDK checkout)"
    );
    return;
  }

  const port = await findFreePort();
  const token = randomToken();
  const logPath = path.join(os.tmpdir(), `kern-live-${process.pid}-${Date.now()}.log`);
  const logFd = openSync(logPath, "w");
  const proc = spawn(cmd[0], [...cmd.slice(1), "--addr", `127.0.0.1:${port}`], {
    cwd: repoRoot,
    env: { ...process.env, KERN_AUTH_TOKEN: token },
    stdio: ["ignore", logFd, logFd],
    detached: true, // own process group: the go-run server child dies with it
  });
  const base = `http://127.0.0.1:${port}`;
  try {
    const ready = await waitForKernServer(base, token, proc, logPath, 240000, {
      usedPrebuilt,
      onSkip: (reason) => t.skip(reason),
      onStop: () => {
        try {
          process.kill(-proc.pid, "SIGTERM");
        } catch {
          /* already gone */
        }
      },
    });
    if (!ready) return;

    // The mocked tests above replaced global.fetch and never restored it,
    // so put the REAL fetch back before touching the live server.
    global.fetch = realFetch;
    const authedFetch = async (url, init) => {
      init = {
        ...init,
        headers: { ...(init.headers || {}), Authorization: `Bearer ${token}` },
      };
      return realFetch(url, init);
    };
    try {
      // The server was started with KERN_AUTH_TOKEN set: a bare client (no
      // Authorization header) must be refused with 401 — proves the gate is on.
      const bare = new Client(base);
      await assert.rejects(
        () => bare.callTool("kern_mask_pii", { text: "x" }),
        (e) => e instanceof KernError && e.status === 401
      );
      // From here the client carries the bearer token on every request, then
      // runs the shared fixture's 4 cases against the REAL route: 2xx returns
      // the {"output": ...} payload, 403/404/500 surface as KernError with the
      // mapped status and the documented error body.
      global.fetch = authedFetch;
      const c = new Client(base);
      for (const tc of loadContractFixture()) {
        await t.test(tc.id, async () => {
          try {
            const out = await c.callTool(tc.name, tc.args);
            assert.equal(typeof out, "string");
            assert.ok(out.length > 0, `empty output payload for ${tc.id}`);
            if (tc.id === "success") assert.ok(out.includes("masked 1 secrets"));
            if (tc.id === "tool_failure_500") assert.ok(out.includes("masked 0 secrets"));
          } catch (e) {
            assert.ok(e instanceof KernError, `expected KernError, got ${e}`);
            assert.ok(
              [403, 404, 500].includes(e.status),
              `unexpected status ${e.status} for ${tc.id}`
            );
            if (tc.id === "denied_403") {
              assert.equal(e.status, 403);
              assert.ok(/tool call denied/.test(e.message));
            }
            if (tc.id === "unknown_404") {
              assert.equal(e.status, 404);
              assert.ok(/unknown tool/.test(e.message));
            }
            if (tc.id === "tool_failure_500") {
              assert.equal(e.status, 500);
              assert.ok(/tool call failed/.test(e.message));
            }
          }
        });
      }
    } finally {
      global.fetch = realFetch;
    }
  } finally {
    // Kill the whole process group: `go run` execs the server as a child,
    // and a plain kill on the go process would orphan the server.
    try {
      process.kill(-proc.pid, "SIGTERM");
    } catch {
      proc.kill("SIGTERM");
    }
    await new Promise((res) => {
      const t = setTimeout(res, 15000); // never hang the suite on a stuck server
      proc.once("exit", () => {
        clearTimeout(t);
        res();
      });
    });
    closeSync(logFd);
  }
});

File: sdk/go/kernsdk/client_test.go
package kernsdk

import (
	"context"
	"encoding/json"
	"errors"
	"io"
	"net/http"
	"net/http/httptest"
	"os"
	"path/filepath"
	"reflect"
	"testing"

	"github.com/JayveerPrajapati/kern/internal/domain"
	"github.com/JayveerPrajapati/kern/internal/web"
)

// contractFixture mirrors sdk/contract/tools_call.json — the ONE shared
// route-shape contract for POST /v1/tools/{name} that every shipped SDK
// client suite (Go, Python, TypeScript) asserts against. A route-shape
// change (envelope, status mapping) must fail all three suites until the
// clients are updated.
type contractFixture struct {
	Version int `json:"_version"`
	Cases   []struct {
		ID       string         `json:"id"`
		Name     string         `json:"name"`
		Args     map[string]any `json:"args"`
		Response struct {
			Status int            `json:"status"`
			Body   map[string]any `json:"body"`
		} `json:"response"`
		ExpectOutput string `json:"expect_output"`
	} `json:"cases"`
}

// loadContractFixture reads the shared contract fixture. go test runs with
// cwd = package dir (sdk/go/kernsdk), so the fixture lives two levels up.
func loadContractFixture(t *testing.T) contractFixture {
	t.Helper()
	raw, err := os.ReadFile(filepath.Join("..", "..", "contract", "tools_call.json"))
	if err != nil {
		t.Fatalf("load shared contract fixture: %v", err)
	}
	var fx contractFixture
	if err := json.Unmarshal(raw, &fx); err != nil {
		t.Fatalf("parse shared contract fixture: %v", err)
	}
	if fx.Version != 1 {
		t.Fatalf("contract fixture _version = %d, want 1", fx.Version)
	}
	if len(fx.Cases) == 0 {
		t.Fatal("contract fixture has no cases")
	}
	return fx
}

// sdkFixture writes a tiny Go module so the web console (backed by the same
// TaskService application services the CLI/MCP use) can analyze it.
func sdkFixture(t *testing.T) string {
	t.Helper()
	root := t.TempDir()
	if err := os.WriteFile(filepath.Join(root, "go.mod"), []byte("module sdkserver\n\ngo 1.21\n"), 0o644); err != nil {
		t.Fatal(err)
	}
	src := "package main\n\n// Greet says hello.\nfunc Greet() string { return \"hi\" }\n\nfunc main() { _ = Greet() }\n"
	if err := os.WriteFile(filepath.Join(root, "app.go"), []byte(src), 0o644); err != nil {
		t.Fatal(err)
	}
	return root
}

// TestGoSDKAgainstControlPlane is the exit gate: the Go SDK drives
// the SAME control-plane application services as the CLI and MCP — through the
// same REST surface the Python/TypeScript SDKs use. A real kern-server
// (internal/web over a fixture) backs the client.
func TestGoSDKAgainstControlPlane(t *testing.T) {
	root := sdkFixture(t)
	app, err := web.New(root)
	if err != nil {
		t.Fatalf("web.New: %v", err)
	}
	srv := httptest.NewServer(app)
	defer srv.Close()

	c := New(srv.URL, nil)
	ctx := context.Background()

	// Health.
	h, err := c.Health(ctx)
	if err != nil {
		t.Fatalf("Health: %v", err)
	}
	if h == nil {
		t.Error("Health returned nil")
	}

	// Analyze → context packet (same services as kern analyze / kern_analyze).
	an, err := c.Analyze(ctx, "Greet")
	if err != nil {
		t.Fatalf("Analyze: %v", err)
	}
	if an == nil {
		t.Error("Analyze returned nil")
	}

	// Plan → implementation plan.
	pl, err := c.Plan(ctx, "Greet")
	if err != nil {
		t.Fatalf("Plan: %v", err)
	}
	if pl == nil {
		t.Error("Plan returned nil")
	}

	// What-if → impact simulation.
	wi, err := c.WhatIf(ctx, "remove_symbol", "Greet", "")
	if err != nil {
		t.Fatalf("WhatIf: %v", err)
	}
	if wi == nil {
		t.Error("WhatIf returned nil")
	}

	// Memory add + list.
	if _, err := c.MemoryAdd(ctx, "go sdk lesson", "lesson", "test", []string{"sdk"}); err != nil {
		t.Fatalf("MemoryAdd: %v", err)
	}
	ms, err := c.MemoryList(ctx)
	if err != nil {
		t.Fatalf("MemoryList: %v", err)
	}
	items, _ := ms["items"].([]any)
	if len(items) == 0 {
		t.Errorf("MemoryList returned no memories after MemoryAdd: %v", ms)
	}

	// Task submit + fetch.
	sub, err := c.TaskSubmit(ctx, "Greet", "code")
	if err != nil {
		t.Fatalf("TaskSubmit: %v", err)
	}
	if sub == nil {
		t.Error("TaskSubmit returned nil")
	}

	// Agents roster.
	agents, err := c.Agents(ctx)
	if err != nil {
		t.Fatalf("Agents: %v", err)
	}
	specs, _ := agents["specialists"].([]any)
	if len(specs) == 0 {
		t.Errorf("Agents returned empty roster: %v", agents)
	}

	// Approvals pending roster (D2).
	ap, err := c.ApprovalsPending(ctx)
	if err != nil {
		t.Fatalf("ApprovalsPending: %v", err)
	}
	if ap == nil {
		t.Error("ApprovalsPending returned nil")
	}

	// Incidents list (D2).
	inc, err := c.Incidents(ctx)
	if err != nil {
		t.Fatalf("Incidents: %v", err)
	}
	if inc == nil {
		t.Error("Incidents returned nil")
	}

	// Single incident by id (D2): create one through the same REST surface,
	// then fetch it back.
	var created map[string]any
	if err := c.do(ctx, http.MethodPost, "/api/incidents", map[string]string{
		"title": "sdk test incident", "severity": "error", "status": "OPEN", "affected_service": "checkout",
	}, &created); err != nil {
		t.Fatalf("create incident: %v", err)
	}
	id, _ := created["ID"].(string)
	if id == "" {
		t.Fatalf("created incident missing ID: %v", created)
	}
	one, err := c.Incident(ctx, id)
	if err != nil {
		t.Fatalf("Incident(%s): %v", id, err)
	}
	if one == nil {
		t.Error("Incident returned nil")
	}

	// Events stream opens and carries at least the SSE headers (D2).
	body, err := c.EventsStream(ctx)
	if err != nil {
		t.Fatalf("EventsStream: %v", err)
	}
	defer func() { _ = body.Close() }()
	if _, err := body.Read(make([]byte, 128)); err != nil && err != io.EOF {
		t.Fatalf("EventsStream read: %v", err)
	}
}

// TestGoSDKReturnsStatusErrors verifies the SDK surfaces non-2xx as errors.
func TestGoSDKReturnsStatusErrors(t *testing.T) {
	root := sdkFixture(t)
	app, err := web.New(root)
	if err != nil {
		t.Fatalf("web.New: %v", err)
	}
	srv := httptest.NewServer(app)
	defer srv.Close()

	c := New(srv.URL, nil)
	// A task that does not exist must yield an error, not a silent nil.
	_, err = c.Task(context.Background(), "task-does-not-exist")
	if err == nil {
		t.Fatal("Task(nonexistent) should error")
	}
}

// sdkToolServer is a minimal web.ToolServer double for the CallTool SDK test:
// the /v1/tools/{name} route delegates to CallToolGoverned, so the test
// asserts the full client → route → dispatch contract. A fake is used because
// internal/web cannot import internal/mcp (mcp → org → enterprise → web cycle)
// and the real governed dispatch is already exercised by the internal/sdk
// catalog tests over the same CallToolGoverned path (web/tools_passthrough_test
// uses the identical pattern).
type sdkToolServer struct {
	out      string
	err      error
	lastName string
	lastArgs map[string]any
}

func (s *sdkToolServer) CallToolGoverned(_ context.Context, name string, args map[string]any) (string, error) {
	s.lastName = name
	s.lastArgs = args
	return s.out, s.err
}

func (s *sdkToolServer) Close() {}

// TestGoSDKCallTool covers the client half of the SDK passthrough
// (POST /v1/tools/{name}): CallTool posts the argument map, decodes the
// {"output": ...} envelope and returns the raw tool text; governed denials
// and unknown tools surface as *Err with the mapped HTTP status.
func TestGoSDKCallTool(t *testing.T) {
	// One shared fake whose behavior is mutated between calls — the factory
	// is consulted once at web.New time, so re-registering per scenario
	// would not affect the already-built App (same pattern as the web
	// tools_passthrough_test).
	fake := &sdkToolServer{out: "masked 1 secrets\n"}
	web.SetToolServerFactory(func(string) web.ToolServer { return fake })

	root := sdkFixture(t)
	app, err := web.New(root)
	if err != nil {
		t.Fatalf("web.New: %v", err)
	}
	srv := httptest.NewServer(app)
	defer srv.Close()

	c := New(srv.URL, nil)
	ctx := context.Background()

	// Success: the raw tool output payload is returned.
	out, err := c.CallTool(ctx, "kern_mask_pii", map[string]any{"text": "token=sk-abc123"})
	if err != nil {
		t.Fatalf("CallTool: %v", err)
	}
	if out != "masked 1 secrets\n" {
		t.Errorf("CallTool output = %q, want the raw tool text", out)
	}

	// Unknown tool: the route maps domain.ErrToolUnknown to 404, which the
	// SDK surfaces as *Err carrying the HTTP status.
	fake.out = ""
	fake.err = domain.ErrToolUnknown
	_, err = c.CallTool(ctx, "kern_no_such_tool", map[string]any{})
	var apiErr *Err
	if !errors.As(err, &apiErr) || apiErr.Status != http.StatusNotFound {
		t.Fatalf("unknown tool error = %v, want *Err with status 404", err)
	}

	// Governed denial: domain.ErrToolDenied maps to 403.
	fake.err = domain.ErrToolDenied
	_, err = c.CallTool(ctx, "kern_mask_pii", map[string]any{"root": "/etc"})
	if !errors.As(err, &apiErr) || apiErr.Status != http.StatusForbidden {
		t.Fatalf("denied tool error = %v, want *Err with status 403", err)
	}
}

// TestGoSDKCallToolContractFixture drives CallTool against the SHARED
// contract fixture (sdk/contract/tools_call.json): success returns the
// {"output": ...} payload and forwards the exact name+args to the route, and
// every error class (403 denied, 404 unknown, 500 generic) surfaces as *Err
// with the mapped HTTP status. The Python and TypeScript suites run the same
// cases against the same file, so any route-shape drift fails all three
// suites until the clients are updated.
func TestGoSDKCallToolContractFixture(t *testing.T) {
	fx := loadContractFixture(t)
	fake := &sdkToolServer{out: ""}
	web.SetToolServerFactory(func(string) web.ToolServer { return fake })

	root := sdkFixture(t)
	app, err := web.New(root)
	if err != nil {
		t.Fatalf("web.New: %v", err)
	}
	srv := httptest.NewServer(app)
	defer srv.Close()

	c := New(srv.URL, nil)
	ctx := context.Background()

	for _, tc := range fx.Cases {
		t.Run(tc.ID, func(t *testing.T) {
			// Program the fake for this case class with the sentinel errors
			// the governed path classifies via errors.Is.
			fake.out = ""
			fake.err = nil
			switch tc.Response.Status {
			case http.StatusOK:
				fake.out = tc.ExpectOutput
			case http.StatusForbidden:
				fake.err = domain.ErrToolDenied
			case http.StatusNotFound:
				fake.err = domain.ErrToolUnknown
			case http.StatusInternalServerError:
				fake.err = errors.New("generic tool failure")
			default:
				t.Fatalf("fixture case %q: unhandled status %d", tc.ID, tc.Response.Status)
			}

			out, err := c.CallTool(ctx, tc.Name, tc.Args)

			// Request shape: the exact name and args must reach the route.
			if fake.lastName != tc.Name || !reflect.DeepEqual(fake.lastArgs, tc.Args) {
				t.Errorf("route received (%q, %v), want (%q, %v)", fake.lastName, fake.lastArgs, tc.Name, tc.Args)
			}

			if tc.Response.Status == http.StatusOK {
				if err != nil {
					t.Fatalf("CallTool(%s) = err %v, want output %q", tc.Name, err, tc.ExpectOutput)
				}
				if out != tc.ExpectOutput {
					t.Errorf("CallTool(%s) output = %q, want %q", tc.Name, out, tc.ExpectOutput)
				}
				return
			}
			var apiErr *Err
			if !errors.As(err, &apiErr) || apiErr.Status != tc.Response.Status {
				t.Fatalf("CallTool(%s) error = %v, want *Err with status %d", tc.Name, err, tc.Response.Status)
			}
		})
	}
}
